crates/e2k-py/src/lib.rs

```diff
@@ -0,0 +1,173 @@
+                    if let Some(Ok(top_p)) = top_p {
+                        strategy.top_p = top_p;
+                    }
+                    if let Some(Ok(temperature)) = temperature {
+                        strategy.temperature = temperature;
+                    }
+
+                    e2k::Strategy::TopP(strategy)
+                }
+            }
+        }
+        _ => {
+            return Err(pyo3::exceptions::PyValueError::new_err(
+                "strategy must be one of 'greedy', 'top_k', 'top_p'",
+            ));
+        }
+    })
+}
+
+#[pyclass]
+struct C2k {
+    inner: std::sync::RwLock<e2k::C2k>,
+}
+
+#[pymethods]
+impl C2k {
+    #[new]
+    #[pyo3(signature = (max_len = 32))]
+    fn new(max_len: usize) -> Self {
+        Self {
+            inner: std::sync::RwLock::new(e2k::C2k::new(max_len)),
+        }
+    }
+
+    #[classmethod]
+    #[pyo3(signature = (model, max_len = 32))]
+    fn with_model(_cls: Bound<'_, PyType>, model: Bound<'_, PyBytes>, max_len: usize) -> Self {
+        let model = model.as_bytes();
+        Self {
+            inner: std::sync::RwLock::new(e2k::C2k::with_model(model, max_len)),
+        }
+    }
+
+    #[pyo3(signature = (strategy, **kwargs))]
+    fn set_decode_strategy(
+        slf: &Bound<'_, Self>,
+        strategy: &str,
+        kwargs: Option<&Bound<'_, PyDict>>,
+    ) -> PyResult<()> {
+        let strategy = extract_strategy(strategy, kwargs)?;
+
+        {
+            let inner = &slf.borrow().inner;
+            let mut inner = inner.write().unwrap();
+
+            inner.set_decode_strategy(strategy);
+        };
+
+        Ok(())
+    }
+
+    fn __call__(&self, src: &str) -> String {
+        self.inner.read().unwrap().infer(src)
+    }
+}
+
+#[pyclass]
+struct P2k {
+    inner: std::sync::RwLock<e2k::P2k>,
+}
+
+#[pymethods]
+impl P2k {
+    #[new]
+    #[pyo3(signature = (max_len = 32))]
+    fn new(max_len: usize) -> Self {
+        Self {
+            inner: std::sync::RwLock::new(e2k::P2k::new(max_len)),
+        }
+    }
+
+    #[classmethod]
+    #[pyo3(signature = (model, max_len = 32))]
+    fn with_model(_cls: Bound<'_, PyType>, model: Bound<'_, PyBytes>, max_len: usize) -> Self {
+        let model = model.as_bytes();
+        Self {
+            inner: std::sync::RwLock::new(e2k::P2k::with_model(model, max_len)),
+        }
+    }
+
+    #[pyo3(signature = (strategy, **kwargs))]
+    fn set_decode_strategy(
+        slf: &Bound<'_, Self>,
+        strategy: &str,
+        kwargs: Option<&Bound<'_, PyDict>>,
+    ) -> PyResult<()> {
+        let strategy = extract_strategy(strategy, kwargs)?;
+
+        {
+            let inner = &slf.borrow().inner;
+            let mut inner = inner.write().unwrap();
+
+            inner.set_decode_strategy(strategy);
+        };
+
+        Ok(())
+    }
+
+    fn __call__(&self, src: &Bound<'_, PyList>) -> PyResult<String> {
+        let src: Vec<String> = src.extract()?;
```

Comment:
- [`Vec<impl FromPyObject<'py>>: FromPyObject<'py>`](https://docs.rs/pyo3/0.23/pyo3/conversion/trait.FromPyObject.html#impl-FromPyObject%3C'py%3E-for-Vec%3CT%3E)
- [`String: FromPyObject<'_>`](https://docs.rs/pyo3/0.23/pyo3/conversion/trait.FromPyObject.html#impl-FromPyObject%3C'_%3E-for-String)

```suggestion
    fn __call__(&self, src: Vec<String>) -> PyResult<String> {
```